concourse/pipelines/monitoring.yml

```diff
@@ -0,0 +1,285 @@
+
+
+            aws ecs update-service \
+              --cluster monitoring \
+              --service "$APPLICATION" \
+              --task-definition "$new_task_definition_arn" \
+              --region "$AWS_REGION"
+
+            echo "Waiting for $APPLICATION ECS service to reach steady state..."
+
+            aws ecs wait services-stable \
+              --cluster monitoring \
+              --services "$APPLICATION" \
+              --region "$AWS_REGION"
+
+            echo "Finished updating $APPLICATION to task definition $new_task_definition_arn."
+
+  - name: configure-grafana
+    plan:
+    - get: govuk-infrastructure
+      passed:
+      - deploy-grafana
+      trigger: true
+    - task: terraform-apply
+      config:
+        inputs:
+        - name: govuk-infrastructure
+          path: src
+        outputs:
+        - name: terraform-outputs
+        params:
+          AWS_REGION: eu-west-1
+          ASSUME_ROLE_ARN: 'arn:aws:iam::430354129336:role/govuk-concourse-deployer'
+        platform: linux
+        image_resource:
+          type: docker-image
+          source:
+            repository: governmentpaas/terraform
+            tag: terraform-0.13.3
+            username: ((docker_hub_username))
+            password: ((docker_hub_authtoken))
+        run:
+          path: sh
+          args:
+          - '-c'
+          - |
+            set -eu
+
+            root_dir=$(pwd)
+            echo "${root_dir}"
+
+            INFRA_DIR="${root_dir}/src/terraform/deployments/monitoring-test/infra"
+            cd ${INFRA_DIR}
+            terraform init -backend-config "role_arn=$ASSUME_ROLE_ARN"
+            grafana_fqdn=$(terraform output grafana_fqdn)
+            echo "Grafana base URL: https://${grafana_fqdn}"
+
+            wait_url() {
+              max_retries=10
+              retries_count=0
+              sleep_period=5
+
+              while true ; do
+                  echo "checking if HTTP status code $1 is 200..."
+                  if wget --timeout 2 --server-response $1 2>&1 >/dev/null | awk '/^  HTTP/{print $2}' | grep -q "200" ; then
+                    echo "HTTP status code for $1 is 200"
+                    return
+                  fi
+
+                  retries_count=$((retries_count+1))
+
+                  if [ $retries_count -ge $max_retries ]; then
+                    echo "final HTTP code is not 200 for $1"
+                    exit 1
+                  fi
+
+                  sleep_duration=$((retries_count * sleep_period))
+                  echo "HTTP code is not 200, sleeping for ${sleep_duration}s before new check..."
+                  sleep ${sleep_duration}
+              done;
+            }
+
+            wait_url https://${grafana_fqdn}
+
+            APP_DIR="${root_dir}/src/terraform/deployments/monitoring-test/grafana/app-config"
+            cd ${APP_DIR}
+
+            terraform_apply() {
+              # sed is used here to filter out the ansi colors in the text stream
```

Comment:
I was about to mention that there's a [machine-readable output format](https://www.terraform.io/docs/internals/json-format.html) (`terraform ... -out $file`) then I saw that you're parsing the error messages rather than the actual output. That's both brittle and unnecessary - all you need to do is retry the `terraform apply` (and the clearing state workaround) if it exited nonzero the first time. No need for all this brittle stuff with sed and tee and temporary files.

Reply:
If it is suitable to run the workaround on all errors, then I can do that.

Reply:
Is it possible to simply run `terraform state rm module.grafana-app-config.grafana_data_source.cloudwatch` before every `terraform apply` of the grafana configuration? Rather than parsing errors and doing retries, we could just remove it from the state every time.

Reply:
`terraform state rm module.grafana-app-config.grafana_data_source.cloudwatch` can only be done if terraform was run before. If in clean-slate, we can't do that.